File: scripts/backfill-meeting-summary-embeddings.mjs
#!/usr/bin/env node
/**
 * Backfill document_metadata.summary_embedding for meetings missing it.
 *
 * Uses text-embedding-3-large (3072 dimensions) to match the embedder worker.
 * Prefer AI Gateway, then direct OpenAI fallback, matching the backend pipeline.
 * Requires env vars: DATABASE_URL, AI_GATEWAY_API_KEY or OPENAI_API_KEY
 *
 * Usage:
 *   node scripts/backfill-meeting-summary-embeddings.mjs [--dry-run] [--limit N] [--batch-size N] [--days N]
 */

import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import postgres from 'postgres';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Auto-load credentials from .env (compliance: non-interactive auth, no hardcoded secrets)
try {
  const envFile = readFileSync(join(__dirname, '../.env'), 'utf-8');
  for (const line of envFile.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eqIdx = trimmed.indexOf('=');
    if (eqIdx < 1) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    let val = trimmed.slice(eqIdx + 1).trim();
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    if (!(key in process.env)) process.env[key] = val;
  }
} catch {
  // .env not found; rely on environment variables already set
}

const DATABASE_URL = process.env.DATABASE_URL || process.env.SUPABASE_DB_URL;
const AI_GATEWAY_API_KEY = process.env.AI_GATEWAY_API_KEY;
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

if (!DATABASE_URL || (!AI_GATEWAY_API_KEY && !OPENAI_API_KEY)) {
  console.error("Missing required env vars: DATABASE_URL/SUPABASE_DB_URL and AI_GATEWAY_API_KEY or OPENAI_API_KEY");
  process.exit(1);
}

const args = process.argv.slice(2);
const DRY_RUN = args.includes("--dry-run");
const LIMIT = parseInt(args[args.indexOf("--limit") + 1]) || 0;
const BATCH_SIZE = parseInt(args[args.indexOf("--batch-size") + 1]) || 20;
const DAYS = parseInt(args[args.indexOf("--days") + 1]) || 365;
const sql = postgres(DATABASE_URL, { max: 1, ssl: "require" });

async function embedTexts(texts) {
  const attempts = [
    AI_GATEWAY_API_KEY
      ? {
          label: "AI Gateway",
          url: "https://ai-gateway.vercel.sh/v1/embeddings",
          key: AI_GATEWAY_API_KEY,
          model: "openai/text-embedding-3-large",
        }
      : null,
    OPENAI_API_KEY
      ? {
          label: "OpenAI direct",
          url: "https://api.openai.com/v1/embeddings",
          key: OPENAI_API_KEY,
          model: "text-embedding-3-large",
        }
      : null,
  ].filter(Boolean);

  const errors = [];
  for (const attempt of attempts) {
    const resp = await fetch(attempt.url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${attempt.key}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: attempt.model,
        dimensions: 3072,
        input: texts,
      }),
    });
    if (!resp.ok) {
      const text = await resp.text();
      errors.push(`${attempt.label}: ${resp.status} ${text.slice(0, 300)}`);
      continue;
    }
    const data = await resp.json();
    if (!Array.isArray(data.data) || data.data.length !== texts.length) {
      errors.push(`${attempt.label}: expected ${texts.length} embeddings, got ${data.data?.length ?? "unknown"}`);
      continue;
    }
    console.log(`  Provider: ${attempt.label}`);
    return data.data.sort((a, b) => a.index - b.index).map((d) => d.embedding);
  }

  throw new Error(`Embedding failed across all providers: ${errors.join(" | ")}`);
}

function buildEmbeddingText(meeting) {
  const parts = [];
  if (meeting.title) parts.push(`Meeting: ${meeting.title}`);
  if (meeting.date) parts.push(`Date: ${meeting.date}`);
  const summary = meeting.summary || meeting.overview || "";
  if (summary) parts.push(summary);
  return parts.join("\n");
}

async function fetchAllCandidates() {
  const rows = await sql`
    select id, title, date, captured_at, created_at, summary, overview
    from public.document_metadata
    where summary_embedding is null
      and (source = 'fireflies'
        or type in ('meeting', 'meeting_transcript')
        or category = 'meeting'
        or fireflies_id is not null)
      and coalesce(captured_at, date, created_at::timestamptz) >= now() - (${DAYS} || ' days')::interval
      and length(coalesce(nullif(summary, ''), nullif(overview, ''), '')) > 0
    order by coalesce(captured_at, date, created_at::timestamptz) desc nulls last
    ${LIMIT ? sql`limit ${LIMIT}` : sql``}
  `;

  return rows.filter((m) => buildEmbeddingText(m).trim());
}

async function main() {
  console.log("=".repeat(60));
  console.log("Meeting Summary Embedding Backfill");
  console.log("=".repeat(60));
  console.log(`Model: text-embedding-3-large (3072 dimensions)`);
  console.log(`Batch size: ${BATCH_SIZE}`);
  console.log(`Window: last ${DAYS} days`);
  if (LIMIT) console.log(`Limit: ${LIMIT}`);
  if (DRY_RUN) console.log("MODE: DRY RUN");
  console.log();

  if (!DRY_RUN) {
    console.log("Checking embedding provider...");
    await embedTexts(["alleato meeting summary backfill provider preflight"]);
    console.log();
  }

  // Materialize all candidate meetings upfront so we iterate a stable set.
  // This prevents infinite loops when persistent failures leave rows NULL
  // and avoids double-counting rows that fail across batches.
  console.log("Fetching candidate meetings...");
  let candidates = await fetchAllCandidates();

  if (candidates.length === 0) {
    console.log("No meetings need summary_embedding backfill!");
    return;
  }

  if (LIMIT) candidates = candidates.slice(0, LIMIT);

  console.log(`Found ${candidates.length} meetings to process`);
  console.log("=".repeat(60));
  console.log();

  let processed = 0;
  let success = 0;
  let failed = 0;
  let batchNum = 0;

  for (let batchStart = 0; batchStart < candidates.length; batchStart += BATCH_SIZE) {
    batchNum++;
    const batch = candidates.slice(batchStart, batchStart + BATCH_SIZE);

    console.log(`Batch ${batchNum}: ${batch.length} meetings`);

    const texts = batch.map(buildEmbeddingText);

    if (DRY_RUN) {
      for (const m of batch.slice(0, 3)) {
        const preview = buildEmbeddingText(m).substring(0, 100);
        console.log(`  [DRY RUN] ${m.id}: ${preview}...`);
      }
      if (batch.length > 3) console.log(`  ... and ${batch.length - 3} more`);
      processed += batch.length;
      success += batch.length;
      continue;
    }

    // Generate embeddings for batch
    console.log(`  Generating ${texts.length} embeddings...`);
    const start = Date.now();
    let embeddings;
    try {
      embeddings = await embedTexts(texts);
      console.log(`  Done in ${((Date.now() - start) / 1000).toFixed(2)}s`);
    } catch (err) {
      console.error(`  ERROR generating embeddings: ${err.message}`);
      // Pause before next batch on OpenAI errors to avoid hot-looping
      failed += batch.length;
      processed += batch.length;
      await new Promise((r) => setTimeout(r, 2000));
      continue;
    }

    // Update each meeting
    for (let i = 0; i < batch.length; i++) {
      try {
        await sql`
          update public.document_metadata
          set summary_embedding = ${JSON.stringify(embeddings[i])}::halfvec
          where id = ${batch[i].id}
        `;
        success++;
      } catch (err) {
        console.error(`  Failed ${batch[i].id}: ${err.message}`);
        failed++;
      }
      processed++;
    }

    console.log(`  Progress: ${processed}/${candidates.length} processed, ${success} success, ${failed} failed`);

    // Brief pause between batches
    if (batchStart + BATCH_SIZE < candidates.length) {
      await new Promise((r) => setTimeout(r, 500));
    }
  }

  console.log();
  console.log("=".repeat(60));
  console.log("Backfill Complete");
  console.log("=".repeat(60));
  console.log(`Processed: ${processed}`);
  console.log(`Success:   ${success}`);
  console.log(`Failed:    ${failed}`);
  console.log("=".repeat(60));

  if (failed > 0) process.exit(1);
}

main().catch((err) => {
  console.error("FATAL:", err);
  process.exit(1);
}).finally(async () => {
  await sql.end({ timeout: 5 }).catch(() => {});
});

File: backend/src/scripts/backfill_meeting_summary_embeddings.py
#!/usr/bin/env python3
"""
Backfill script for document_metadata.summary_embedding

Generates embeddings for meetings that have a summary/overview but are missing
their summary_embedding. Uses text-embedding-3-large (3072 dimensions) to match
the embedder worker.

Usage:
    python src/scripts/backfill_meeting_summary_embeddings.py [--batch-size 20] [--limit 100] [--dry-run]

Options:
    --batch-size    Number of meetings to process per batch (default: 20)
    --limit         Maximum number of meetings to process (default: all)
    --dry-run       Show what would be done without making changes
"""

import argparse
import os
import sys
import time
from pathlib import Path
from typing import List, Dict, Any, Optional

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

# Load env from root directory
from dotenv import load_dotenv
root_dir = Path(__file__).parent.parent.parent.parent
env_path = root_dir / '.env'
if not env_path.exists():
    env_path = root_dir / '.env.local'
load_dotenv(env_path)

from openai import OpenAI
from supabase import create_client, Client

EMBEDDING_MODEL = "text-embedding-3-large"
EMBEDDING_DIMENSIONS = 3072


def get_supabase_client() -> Client:
    url = os.getenv("SUPABASE_URL")
    key = os.getenv("SUPABASE_SERVICE_ROLE_KEY") or os.getenv("SUPABASE_SERVICE_KEY")
    if not url or not key:
        raise RuntimeError("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY are required")
    return create_client(url, key)


def get_openai_client() -> OpenAI:
    api_key = os.getenv("OPENAI_API_KEY")
    if not api_key:
        raise RuntimeError("OPENAI_API_KEY is required")
    return OpenAI(api_key=api_key)


def get_embedding_text(meeting: Dict[str, Any]) -> str:
    """Build canonical text to embed from meeting fields.

    Uses the same format as the embedder worker and Node.js backfill:
    'Meeting: {title}\\nDate: {date}\\n{summary || overview}'
    """
    title = meeting.get('title') or ''
    date = meeting.get('date') or ''
    summary = meeting.get('summary') or meeting.get('overview') or ''

    parts = []
    if title:
        parts.append(f"Meeting: {title}")
    if date:
        parts.append(f"Date: {date}")
    if summary:
        parts.append(summary)

    return "\n".join(parts)


def fetch_all_candidates(
    supabase: Client,
    limit: Optional[int] = None,
) -> List[Dict[str, Any]]:
    """Fetch all candidate meetings upfront as a stable snapshot.

    Fetches all meetings with NULL summary_embedding that have embeddable text.
    Processing a stable ID list (rather than re-querying each batch) ensures no
    meetings are skipped and failed rows are not double-counted.
    """
    PAGE_SIZE = 1000
    all_candidates = []
    offset = 0

    while True:
        result = (
            supabase.table('document_metadata')
            .select('id, title, date, summary, overview')
            .is_('summary_embedding', 'null')
            .order('date', desc=True)
            .range(offset, offset + PAGE_SIZE - 1)
            .execute()
        )
        rows = result.data or []
        # Keep only rows that have text to embed
        embeddable = [
            r
            for r in rows
            if get_embedding_text(r).strip()
            and "interview" not in str(r.get("title") or "").lower()
        ]
        all_candidates.extend(embeddable)

        if limit and len(all_candidates) >= limit:
            all_candidates = all_candidates[:limit]
            break
        if len(rows) < PAGE_SIZE:
            break
        offset += PAGE_SIZE

    return all_candidates


def generate_embeddings(
    openai_client: OpenAI,
    texts: List[str],
) -> List[List[float]]:
    """Generate 3072-dim embeddings for a list of texts."""
    if not texts:
        return []

    response = openai_client.embeddings.create(
        model=EMBEDDING_MODEL,
        dimensions=EMBEDDING_DIMENSIONS,
        input=texts,
    )

    result: List[List[float]] = [[] for _ in texts]
    for item in response.data:
        result[item.index] = item.embedding

    return result


def backfill(
    batch_size: int = 20,
    limit: Optional[int] = None,
    dry_run: bool = False,
) -> Dict[str, int]:
    supabase = get_supabase_client()
    openai_client = get_openai_client()

    print(f"\n{'='*60}")
    print(f"Meeting Summary Embedding Backfill")
    print(f"{'='*60}")
    print(f"Batch size: {batch_size}")
    print(f"Model: {EMBEDDING_MODEL} ({EMBEDDING_DIMENSIONS} dimensions)")
    if limit:
        print(f"Limit: {limit}")
    if dry_run:
        print(f"MODE: DRY RUN (no changes will be made)")
    print()

    # Materialize all candidates upfront for a stable processing set.
    # This prevents skipping rows that fail (they stay NULL and would be re-fetched
    # indefinitely with offset=0), and avoids double-counting across batches.
    print("Fetching candidate meetings...")
    candidates = fetch_all_candidates(supabase, limit=limit)

    print(f"Total meetings to process: {len(candidates)}")
    print(f"{'='*60}\n")

    if not candidates:
        print("No meetings need summary_embedding backfill!")
        return {'processed': 0, 'success': 0, 'failed': 0}

    stats = {'processed': 0, 'success': 0, 'failed': 0}
    total = len(candidates)
    batch_num = 0

    for batch_start in range(0, total, batch_size):
        batch_num += 1
        batch = candidates[batch_start:batch_start + batch_size]

        print(f"Batch {batch_num}: {len(batch)} meetings")
        texts = [get_embedding_text(m) for m in batch]

        if dry_run:
            print(f"  [DRY RUN] Would generate {len(texts)} embeddings")
            for m, text in zip(batch[:3], texts[:3]):
                preview = text[:100] + "..." if len(text) > 100 else text
                print(f"    - {m['id']}: {preview}")
            if len(batch) > 3:
                print(f"    ... and {len(batch) - 3} more")
            stats['processed'] += len(batch)
            stats['success'] += len(batch)
            continue

        # Generate embeddings
        print(f"  Generating {len(texts)} embeddings...")
        start_time = time.time()
        try:
            embeddings = generate_embeddings(openai_client, texts)
            elapsed = time.time() - start_time
            print(f"  Embeddings generated in {elapsed:.2f}s")
        except Exception as e:
            print(f"  ERROR generating embeddings: {e}")
            stats['failed'] += len(batch)
            stats['processed'] += len(batch)
            # Pause before retrying next batch on API errors
            time.sleep(2.0)
            continue

        # Update database for each meeting in batch
        for meeting, embedding in zip(batch, embeddings):
            if not embedding:
                stats['failed'] += 1
                stats['processed'] += 1
                continue
            try:
                supabase.table('document_metadata').update({
                    'summary_embedding': embedding,
                }).eq('id', meeting['id']).execute()
                stats['success'] += 1
            except Exception as e:
                print(f"  Error updating {meeting['id']}: {e}")
                stats['failed'] += 1
            stats['processed'] += 1

        pct = 100 * stats['processed'] / total
        print(f"  Batch {batch_num} complete: {stats['success']} success, {stats['failed']} failed")
        print(f"  Progress: {stats['processed']}/{total} ({pct:.1f}%)")

        if batch_start + batch_size < total:
            time.sleep(0.5)

    print(f"\n{'='*60}")
    print(f"Backfill Complete")
    print(f"{'='*60}")
    print(f"Processed: {stats['processed']}")
    print(f"Success:   {stats['success']}")
    print(f"Failed:    {stats['failed']}")
    print(f"{'='*60}\n")

    return stats


def main():
    parser = argparse.ArgumentParser(
        description="Backfill document_metadata.summary_embedding with OpenAI embeddings"
    )
    parser.add_argument('--batch-size', type=int, default=20)
    parser.add_argument('--limit', type=int, default=None)
    parser.add_argument('--dry-run', action='store_true')

    args = parser.parse_args()

    try:
        stats = backfill(
            batch_size=args.batch_size,
            limit=args.limit,
            dry_run=args.dry_run,
        )
        if stats['failed'] > 0:
            sys.exit(1)
    except KeyboardInterrupt:
        print("\n\nInterrupted by user")
        sys.exit(130)
    except Exception as e:
        print(f"\nFATAL ERROR: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()
